prometheus/go_collector_go117_test.go

```diff
@@ -0,0 +1,222 @@
+	goMetrics := collectGoMetrics(t)
+
+	var mhist Metric
+	for _, m := range goMetrics {
+		if m.Desc().fqName == "go_gc_heap_allocs_by_size_bytes_total" {
+			mhist = m
+			break
+		}
+	}
+	if mhist == nil {
+		t.Fatal("failed to find metric to test")
+	}
+	hist, ok := mhist.(*constRuntimeMetricsHistogram)
+	if !ok {
+		t.Fatal("found metric is not a runtime/metrics histogram")
+	}
+
+	// Make a bunch of allocations then do another collection.
+	//
+	// The runtime/metrics API tries to reuse memory where possible,
+	// so make sure that we didn't hang on to any of that memory in
+	// hist.
+	countsCopy := make([]uint64, len(hist.counts))
+	copy(countsCopy, hist.counts)
+	for i := 0; i < 100; i++ {
+		sink = make([]byte, 128)
+	}
+	collectGoMetrics(t)
+	for i, v := range hist.counts {
+		if v != countsCopy[i] {
+			t.Error("counts changed during new collection")
+			break
+		}
+	}
+
+	// Get the runtime/metrics copy.
+	s := []metrics.Sample{
+		{Name: "/gc/heap/allocs-by-size:bytes"},
+	}
+	metrics.Read(s)
+	rmHist := s[0].Value.Float64Histogram()
+	// runtime/metrics histograms always have -Inf and +Inf buckets.
+	// We never handle -Inf and +Inf is implicit.
+	wantBuckets := len(rmHist.Buckets) - 2
+
+	// Check to make sure the output proto makes sense.
+	pb := &dto.Metric{}
+	hist.Write(pb)
+
+	if math.IsInf(pb.Histogram.Bucket[len(pb.Histogram.Bucket)-1].GetUpperBound(), +1) {
+		t.Errorf("found +Inf bucket")
+	}
+	if got := len(pb.Histogram.Bucket); got != wantBuckets {
+		t.Errorf("got %d buckets in protobuf, want %d", got, wantBuckets)
+	}
+	for i, bucket := range pb.Histogram.Bucket {
+		// runtime/metrics histograms are lower-bound inclusive, but we're
+		// upper-bound inclusive. So just make sure the new inclusive upper
+		// bound is somewhere close by (in some cases it's equal).
+		wantBound := rmHist.Buckets[i+1]
+		if gotBound := *bucket.UpperBound; (wantBound-gotBound)/wantBound > 0.001 {
+			t.Errorf("got bound %f, want within 0.1%% of %f", gotBound, wantBound)
+		}
+		// Make sure counts are cumulative. Because of the consistency guarantees
+		// made by the runtime/metrics package, we're really not guaranteed to get
+		// anything even remotely the same here.
+		if i > 0 && *bucket.CumulativeCount < *pb.Histogram.Bucket[i-1].CumulativeCount {
+			t.Error("cumulative counts are non-monotonic")
+		}
+	}
+}
+
+func collectGoMetrics(t *testing.T) []Metric {
+	t.Helper()
+
+	c := NewGoCollector().(*goCollector)
+
+	// Collect all metrics.
+	ch := make(chan Metric)
+	var wg sync.WaitGroup
+	var metrics []Metric
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for metric := range ch {
+			metrics = append(metrics, metric)
+		}
+	}()
+	c.Collect(ch)
+	close(ch)
+
+	wg.Wait()
+
+	return metrics
+}
+
+func TestNameFromRuntimeMetrics(t *testing.T) {
+	tests := []struct {
+		got    metrics.Description
+		expect string
+	}{
+		{metrics.Description{Name: "/memory/live:bytes"}, "go_memory_live_bytes"},
```

Comment:
Should we check all metrics to ensure we know what change in terms of metric naming of those across Go versions? Essentially it's quite scary if Go suddenly changes the internal naming of one important metrics e.g go_memory_allocs_bytes_total which makes our code to produce e.g `go_memory_alloc_bytes_total` instead. This will break many setups in a hidden way.

Reply:
I elaborated on this on the main comment thread, but I'll note again for posterity:
1. the Go team takes backwards compatibility very seriously, and metric names effectively *never* change meaning (see the documentation of the package https://pkg.go.dev/runtime/metrics). new metrics will be added with some regularity, and old ones eventually removed as they no longer become useful (less regularly). they will always be announced ahead of time, and will happen at a max speed of once every six months (far, far less than that in practice).
2. I added a test that, for a given Go release, checks for the set of expected metrics. the new set of expected metrics for a Go release can be regenerated with a `go generate` script I added to the source tree. this way, at least to the Prometheus team, it reduces the risk of anything breaking silently and differences may be more easily audited.

do you feel like this alleviates your concerns? I'm happy to discuss more.

Reply:
With a test, I think I am happy, commented more above 🤗 Thanks!